gnocchi/rest/api.py

```diff
@@ -1484,20 +1482,33 @@ def post(self, metric_id, start=None, stop=None, aggregation='mean',
                 abort(400, "Invalid value for stop")
 
         try:
-            return {
-                str(metric.id): values
-                for metric, values in six.iteritems(
-                    pecan.request.storage.search_value(
-                        metrics, query, start, stop, aggregation,
-                        granularity
-                    )
-                )
-            }
+            predicate = storage.MeasureQuery(query)
         except storage.InvalidQuery as e:
             abort(400, six.text_type(e))
+
+        if granularity is not None:
+            granularity = [utils.to_timespan(g)
+                           for g in arg_to_list(granularity)]
+
+        results = {}
+
+        try:
+            for metric in metrics:
+                if granularity is None:
+                    granularity = [d.granularity
+                                   for d in metric.archive_policy.definition]
+                results[str(metric.id)] = []
+                for r in utils.parallel_map(
+                        pecan.request.storage.find_measure,
+                        ((metric, predicate, g, aggregation,
+                          start, stop)
+                         for g in sorted(granularity))):
```

Comment:
we're needlessly sorting repeatedly if granularity is provided. and it's done in multithread so it's not guaranteed to do return sorted.

also in original code it's sorting in descending order.

Reply:
> and it's done in multithread so it's not guaranteed to do return sorted

It's not guaranteed to be executed in order, but it's returned ordered.